Port gwd_compute_tendencies_from_stress_divergence to CXX

jgfouca opened the pull request:
Change list:
1) Main change: Fill out CXX impl of gwd_compute_tendencies_from_stress_divergence
2) gw_common_init: Allow it to be called multiple times be deallocating cref and alpha if
   they had been previously allocated.
3) Begin to fill in constants needed by GW
4) Create struct for containing common init data in C++
5) Add C++ version of gw_common_init and gw_common_finalize
6) Don't just assume all C++ data is packed. Just have it all unpacked for now
7) Fix randomized test data for gwd_compute_tendencies_from_stress_divergence so
   that some actual interesting computations happen.
8) For now, have a f90 and cxx bridge available for this function and use the f90
   one for baseline generation. Once we are sure we have things working, the f90
   calls can be removed.

Discussion items:
* The current CXX is an exact 1:1 port of the fortran, which is not what we want long term
* Like in p3, should we assume these GW functions are called from kernel that is already parallelizing over columns? So, any fortran array like a 3d (ncol, X, Y) becomes a CXX 2d view (X, Y) because we will be passing in subviews for specific columns?
* Should we assume all vertical column data is packed?
* Since I wasn't 100% sure about where these GW will sit in the parallel hierarchy, the C++ impl is fully serial right now which will obviously need to be changed.

bartgol reviewed (commented):
I don't know anything about GW, so I will let other do a more menaingful review.

As for the question of whether a fcn should expect 2d vs 1d views, and whether it should expect Real or Pack data:
- isn't GW a single-col model? If so, why not implementing ALL fcn with 1d args, and then we can wrap the col loop from the eamxx process level (or from another function), leaving the door open for a SK approach like in p3/shoc?
- I don't know how possible this is, but I would encourage using agnostic scalar type as much as possible. I saw a few k+1 deps, so in certain places that would be more delicate, but there are a few tools to compute, e.g., consecutive levs differences, which work with both Real and Pack types.

ambrad commented on components/eamxx/src/physics/gw/impl/gw_gw_common_init_impl.hpp:
```diff
@@ -0,0 +1,55 @@
+ */
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
```
I think this won't be called from a kernel. This is a top-level, one-time init.

ambrad commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -20,27 +27,150 @@ void Functions<S,D>::gwd_compute_tendencies_from_stress_divergence(
-  const uview_1d<const Spack>& nm,
-  const uview_1d<const Spack>& xv,
-  const uview_1d<const Spack>& yv,
+  const uview_1d<const Real>& lat,
```
It's fine to go with Real here if you've decided that at the top level you'll take all views from the AD as Real arrays rather than pack arrays.

jgfouca replied:
I haven't decided that. I was hoping someone would know for sure if we want to do that or not. Using `Real` definitely makes the code easier.

ambrad replied:
My opinion is that it's fine for individual parameterizations to decide not to use packs.

ambrad commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -11,6 +12,12 @@ namespace gw {
+}
+
 template<typename S, typename D>
 KOKKOS_FUNCTION
```
Here again I'm thinking this is not a function called from a kernel, since there are view allocs below and loops over all the cols. So `KOKKOS_FUNCTION` shouldn't be used.

ambrad commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -20,27 +27,150 @@ void Functions<S,D>::gwd_compute_tendencies_from_stress_divergence(
+
+      // Determine the wind tendency, including excess stress carried down
+      // from above.
+      for (int i = 0; i < ncol; ++i) {
```
I suspect you'll want to put the `k` loop inside multiple `ncol` loops and parallelize as in P3 and SHOC: top-level TeamPolicy parallel_for over columns, then team parallelism over levels.

jgfouca replied:
I agree. The question is, will the top-level col kernel already be launched before this function is called? That's what we did in p3.

jgfouca replied:
Yes, Luca mentioned that. I think we are all in agreement then, thx!

jgfouca commented:
Update: I fixed all the suggestions from earlier and I got it working on GPU. There were some complexities that came up:
1) The main k-lev loop has to be serialized because the k+1 computations depend on k.
2) Due to (1), I changed it so that the parallel loop is over -ngwv:ngwv, since each wave computation is independent of the others.
3) (2) made it more complicated to get the ubt sum in a repeatable fashion. I ended up storing all the contributions in a work view `work(ngwv*2 + 1, plev)` and then doing the sum in serial.

ambrad commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -15,32 +17,121 @@ template<typename S, typename D>
+      tau(i,0) = 0.;
+    }
+  }
+
```
I think you need a barrier here before reading tau below.

Alternatively, since the loop below is || is ngwv but serial in k, I'm thinking this 0'ing go right before the serial for loop on k.

jgfouca replied:
The main loop is over -ngwv:ngwv, but the tau's first dimension is -pgwv:pgwv, so I don't know if I can do it the way you are suggesting. I can put this 0-ing in a small kernel to reduce races.

ambrad replied:
Inspection of the F90 makes me believe ngwv is an argument to subroutines, and the callers pass pgwv as the value of ngwv. Thus, I provisionally concluded ngwv=pgwv.

ambrad replied:
That's another good reason to take some time to determine this. You could do code inspection to see if my code inspection deduced the right thing. You could also ping Walter Hannah to find out what he thinks. Being able to drop one of the two index ranges would be nice.

jgfouca replied:
Done

jgfouca commented on components/eamxx/src/share/util/eamxx_utils.hpp:
```diff
@@ -71,6 +71,13 @@ void sort (std::list<T>& l) {
+// Provides functionality identical to fortran's SIGN function
+template <typename T>
+KOKKOS_INLINE_FUNCTION
+T sign(T a, T b) {
```
@mahf708 , yes, you mentioned that earlier and I forgot to follow up on it, sry.

Yes, that looks like it exactly matches fortran's SIGN. The issue is, is it callable from the device?

jgfouca replied:
Nice! I will use that instead.

bartgol commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -15,32 +17,124 @@ template<typename S, typename D>
+      if (k <= tend_level) {
+        // Save tendency for each wave (for later computation of kzz),
+        // applying efficiency and taper:
+        gwut(k,nl_idx) = Kokkos::copysign(ubtl, c(pl_idx)-ubm(k)) * effgw * ptaper;
```
Nice, didn't know kokkos had this fcn.

jgfouca commented:
Based on inspection and feedback from Andrew and Walter, I'm confident we can remove the ngwv concept. I will remove it now.

jgfouca commented:
Baseline diffs in the gw tests are expected since I re-worked the init data for all the tests.

bartgol commented on components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp:
```diff
@@ -47,22 +46,16 @@ void Functions<S,D>::gwd_compute_tendencies_from_stress_divergence(
-    int pl_idx = nl_idx + (pgwv - ngwv); // 0-based idx -pgwv:pgwv arrays
+    int pl_idx = l + pgwv; // 0-based idx for -pgwv:pgwv arrays
+
+    // Force tau at the top of the model to zero, if requested.
```
I _think_ this will cause compute-sanitizer to issue an error for the racecheck condition. I'm not sure, I don't know how it handles concurrent writes with the same compite-time number, maybe it recognizes that it's benign?
